[GSOC23] - C - Implement a StAX parser for OVAL files

HoussemNasri opened the pull request:
## What does this PR change?

Introduce a StAX parser (instead of the current JAXB parser) for OVAL files to increase memory efficiency when parsing large OVAL files.

### Useful Links
- [The RFC](https://github.com/uyuni-project/uyuni-rfc/pull/80)
## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: **Only backend changes**
- [x] **DONE**

## Test coverage
- No tests: **test adapted**

- [x] **DONE**

## Links

Fixes https://github.com/SUSE/spacewalk/issues/26814

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)


## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests" 
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

HoussemNasri committed:
Parse OVAL states and objects with StAX

# Conflicts:
#	java/core/src/main/java/com/suse/oval/OvalParser.java

HoussemNasri committed:
Parse OVAL tests with StAX

HoussemNasri committed:
Parse OVAL definition's metadata with StAX

HoussemNasri committed:
Parse OVAL definition's criteria tree with StAX

HoussemNasri committed:
Support parsing Dpkg OVAL with StAX

HoussemNasri committed:
Rename 'OVALLookupHelper' to 'OVALResourceCache'

 Conflicts:
	java/code/src/com/suse/oval/manager/OVALLookupHelper.java
	java/core/src/main/java/com/suse/oval/manager/OVALLookupHelper.java
	java/core/src/main/java/com/suse/oval/manager/OVALResourcesCache.java

mcalmer commented:
@HoussemNasri @parlt91 i think this is still needed, right?

HoussemNasri commented:
> @HoussemNasri @parlt91 i think this is still needed, right?

Yes, the A and B pull requests need to be reviewed first before we can move to this one in case there was some changes that would affect this one (all PRs are kind of stacked on each other). Right now, pr A is under review.

mcalmer commented:
Than we better reopen this request

mcalmer commented:
@HoussemNasri @parlt91 if this is still needed, please say so or reopen

HoussemNasri committed:
Parse and sync OVAL definitions in bulk

# Conflicts:
#	java/code/src/com/suse/oval/OvalParser.java
#	java/code/src/com/suse/oval/parser/OvalParser.java
#	java/core/src/main/java/com/redhat/rhn/manager/audit/CVEAuditManagerOVAL.java
#	java/core/src/main/java/com/suse/oval/OvalParser.java

Etheryte commented on web/html/src/branding/css/susemanager/components/cveaudit.less:
```diff
@@ -0,0 +1,10 @@
+.scan-data-indicator {
```
This file doesn't belong here, the right place to put these styles is in a style module that's adjacent to the view where you use it. In this specific case for example `web/html/src/manager/audit/cveaudit/cveaudit.module.less`. This also requires changes to how you use the classname in code. You can see `web/html/src/components/table/SearchPanel.module.less` and `web/html/src/components/table/SearchPanel.tsx` for an example on how to use style modules.

The components folder is for very generic styles for things such as inputs, buttons, etc, not for view-specific styles.

HoussemNasri replied:
I messed something up while resolving conflicts but it should be removed.

Etheryte replied:
Okay, thanks for clarifying. In that case it might be good to double check if there's anything else that might've gone missing in this process. One option to avoid issues like this for long-running brances would be to squash the change sets once a certain part gets approved. This would make rebasing easier, but on the other hand reduce the amount of history that's available, so it's mostly a question of whether the trade off is worth it for you.

Etheryte reviewed (changes requested):
The issues that were pointed out regarding styles and how they're used in the previous pull request are still present in this one. The comments are marked as resolved, but the issues have not been addressed, see e.g. https://github.com/uyuni-project/uyuni/pull/7509#discussion_r1510919310 and https://github.com/uyuni-project/uyuni/pull/7509#discussion_r1510923643. Please do not mark issues as resolved if they have not been addressed.

Etheryte reviewed (dismissed):
No frontend changes in the latest summa summarum changeset.

HoussemNasri commented on java/code/src/com/suse/oval/OVALCleaner.java:
```diff
@@ -82,6 +81,7 @@ private static void doCleanupDefinition(DefinitionType definition, OsFamily osFa
 
         if (osFamily == OsFamily.DEBIAN) {
             convertDebianTestRefs(definition.getCriteria(), osVersion);
+            // SUSE-MicroOS-release is ==5.3
```
Hmmm, it should work with LeapMicro and SUSE Micro as far as I remember, unless they have changed something in their schema.

HoussemNasri replied:
Done.

parlt91 reviewed (changes requested):
I also did some testing around the changes. There seems to be an issue with parsing. A lot of info seems to be overridden.  If I had to guess I'd say it looks like each bulk overrides the last one instead of adding onto it. Could you have a look into it?
I'll do another review round once these issues are addressed.

HoussemNasri committed:
Introduce suseOVALOsProduct table

- Also referenced the table from suseOVALPlatformVulnerablePackage

 Conflicts:
	schema/spacewalk/common/tables/tables.deps
	schema/spacewalk/upgrade/susemanager-schema-5.2.0-to-susemanager-schema-5.2.1/300-remove-scap-filesize-limit-from-db.sql

HoussemNasri committed:
Specify the OS product to which the OVAL vulnerability entry belongs in the database

HoussemNasri committed:
Rename 'OVALProduct' to 'OVALOsProduct'

HoussemNasri committed:
Clear existing OVAL metadata correctly before inserting new data for a given OS product

# Conflicts:
#	java/code/src/com/redhat/rhn/common/db/datasource/xml/oval_queries.xml
#	java/core/src/main/java/com/suse/oval/OVALCachingFactory.java

HoussemNasri committed:
Remove redundant comment

HoussemNasri committed:
Use OVAL Stax parser in tests

HoussemNasri committed:
Update copyright year to 2024

parlt91 reviewed (commented):
Could you go a bit into detail on your though process for adding the `ovalOsProduct` table? IIUC you use it to determine what data we need to delete from `ovalPlatfromVulnerablePackage` on resync of the oval data to avoid conflicting data.

HoussemNasri commented:
> IIUC you use it to determine what data we need to delete from ovalPlatfromVulnerablePackage on resync of the oval data to avoid conflicting data.

Correct.

An OVAL product is the abstraction we use in syncing OVAL metadata. It is the combination of OS and version as configured in the `oval.config.json`. I tried using the exisitng OVALPlatform table but in the current design we cannot sync each OVAL platform individually, thus we need the extra table.

HoussemNasri commented on java/code/src/com/suse/oval/OVALCachingFactory.java:
```diff
@@ -73,17 +74,19 @@ public static void savePlatformsVulnerablePackages(OvalRootType rootType) {
-        productVulnerablePackages.stream()
-                .collect(groupingBy(ProductVulnerablePackages::getProductCpe))
-                .keySet().forEach(OVALCachingFactory::clearOVALMetadataByPlatform);
+//        productVulnerablePackages.stream()
```
Outdated, was overridden by a later commit.

HoussemNasri committed:
Switch to cascading delete for OVAL metadata cleanup

HoussemNasri commented on java/code/src/com/redhat/rhn/common/db/datasource/xml/oval_queries.xml:
```diff
@@ -46,10 +46,18 @@
+        <query params="os_product_family, os_product_version">
+            DELETE
+            FROM suseOVALPlatformVulnerablePackage pvp
+            WHERE pvp.platform_id = (SELECT id
```
Didn't know we can do that, thanks for the suggestion.

HoussemNasri commented on schema/spacewalk/upgrade/susemanager-schema-5.1.1-to-susemanager-schema-5.1.2/003-map-oval-vulnerability-metadata-by-os-product.sql:
```diff
@@ -0,0 +1,108 @@
+    DROP CONSTRAINT suse_oval_platform_vulnerable_pkg_id_pk;
+ALTER TABLE suseOVALPlatformVulnerablePackage
+    ADD CONSTRAINT suse_oval_platform_vulnerable_pkg_id_pk PRIMARY KEY (product_os_id, platform_id, cve_id, vulnerable_pkg_id);
+
```
Do we need to add it to the table definition?

HoussemNasri committed:
Fix typo in comment

admd commented:
@HoussemNasri it would be great if you could look into Pascal's comments and address them, so we could merge the remaining PRs and be done with it.

Thank you for all the work, we have only few last bits pending and it would be amazing if we could finish it up in coming weeks.

HoussemNasri committed:
Add ON DELETE CASCADE to suseOVALPlatformVulnerablePackage.suseOVALOsProduct foreign key

HoussemNasri committed:
Ensure SQL migration script in idempotent

 Conflicts:
	schema/spacewalk/upgrade/susemanager-schema-5.1.1-to-susemanager-schema-5.1.2/003-map-oval-vulnerability-metadata-by-os-product.sql

HoussemNasri commented on java/code/src/com/suse/oval/vulnerablepkgextractor/SUSEVulnerablePackageExtractor.java:
```diff
@@ -121,7 +121,7 @@ else if (comment.endsWith("is not affected")) {
             String comment = productCriterion.getComment();
             String productUserFriendlyName = comment.replace(" is installed", "");
-            TestType productTest = ovalLookupHelper.lookupTestById(productCriterion.getTestRef()).orElseThrow();
+            TestType productTest = ovalResourcesCache.lookupTestById(productCriterion.getTestRef()).orElseThrow();
```
There is a bug in the XML; it references an OVAL test that lacks a corresponding `rpminfo_test` in the same OVAL file. We can recommend that the team responsible for OVAL data production add an additional validation step to their pipeline to check references, but I have no idea how to reach them.


You are right about your question; we need to be as resilient as possible in the face of errors otherwise the feature would keep being broken most of the time.

admd replied:
@msmeissn is this something feasible for SUSE provided oval files?^^

HoussemNasri replied:
As I said to Abid and Pascal, this turned out not to be a bug in the XML but rather an unsupported feature from our side. We currently only collect OVAL tests defined with `rpminfo_test` but there are other ways to define tests that I was not aware of, like the one below:

```xml
<unix-def:uname_test xmlns:unix-def="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" check="at least one" comment="kernel version 4.4.*TDC is currently running" id="oval:org.opensuse.security:tst:2009876317" version="1">
		<unix-def:object object_ref="oval:org.opensuse.security:obj:2009060496"/>
		<unix-def:state state_ref="oval:org.opensuse.security:ste:2009206404"/>
</unix-def:uname_test>
```

Supporting `uname_test` would require implementing a general-purpose OVAL scanner, the one we have is a special purpose, in that it decides the vulnerability state of a machine based on whether it has vulnerable RPM/DPKG packages installed or not which is not always true nor sufficient (.e.g. the machine is only vulnerable if the vulnerable package is configured a certain way), but it works most of the time. The best we can do is to log and ignore whenever we detect the usage of uname_test.

HoussemNasri committed:
Be more resilient in face of errors

HoussemNasri committed:
Resolve potential XML XXE vulnerability

HoussemNasri committed:
Add changelogs

HoussemNasri committed:
Fix trailing whitespace in changelog

HoussemNasri committed:
Fix some sonar warnings

HoussemNasri committed:
Use stax2-api

admd commented:
> > Look at failing SonarCloud tests
> 
> Done.
> 
> > Look into the comment I left in gitter earlier about replacing stax-api-1. Can you check if it can be replaced by stax2-api provided by org/codehaus/stax2 without to much effort?
> 
> We still use some `javax.xml` classes in OVALParser, not sure how to get rid of them or if it's possible.

@HoussemNasri I don't think the idea is to replace all `javax.xml` classes. Some of them belongs to JAXB and other java XML APIs which we are still keeping it. We are only interested in not having javax.xml.stream.*` classes which belong to stax-api-1.0.1. I see you are already using now stax2-api in ovalParser at some places and only few classes are left from stax-api-1.0.1 which needs to be replaced with corresponding classes from stax2-api

parlt91 committed:
Fix sonarcloud issues

Signed-off-by: Pascal Arlt <parlt@suse.com>

parlt91 committed:
Move schema migration file

Signed-off-by: Pascal Arlt <parlt@suse.com>

rjmateus commented on java/code/src/com/suse/oval/parser/OvalParser.java:
```diff
@@ -0,0 +1,724 @@
+     * @param ovalFileURL the OVAL file to parse
+     * @return the parsed OVAL encapsulated in an {@link OvalRootType} object.
+     * */
+    public OvalRootType parse(URL ovalFileURL) throws OvalParserException {
```
This parser is only used for testing. Should we move it to a test-only class? The method could be a utility that would receive a `OvalParser` instance and the File location. Since it keeps all in memory, I think it is risky to have in the main class when is used only for testing.

rjmateus commented on java/code/src/com/suse/oval/parser/OvalParser.java:
```diff
@@ -0,0 +1,724 @@
+     * @param ovalFile an XML file containing OVAL definitions to be parsed.
+     * @return all OVAL definitions in {@code ovalFile}
+     * */
+    public List<DefinitionType> parseAllDefinitions(File ovalFile) {
```
This is a dependency for the parser function that is used only for testing. We want people to use the bulk parser, and not the one that loads everything to memory.

rjmateus commented on java/code/src/com/redhat/rhn/manager/audit/CVEAuditManagerOVAL.java:
```diff
@@ -418,24 +423,33 @@ private static void syncOVALForProduct(OVALProduct product, OVALDownloader ovalD
+        OvalParser ovalParser = new OvalParser();
+        OVALResources ovalResources = ovalParser.parseResources(ovalFile);
+        ovalParser.parseDefinitionsInBulk(ovalFile, definitionsBulk -> {
+            OvalRootType ovalRoot = new OvalRootType();
```
Looks to me that the `product.getOsFamily()` and `product.getOsVersion()` should also be use to set the `OvalRootType`.
In the method called `savePlatformsVulnerablePackages` the OS family and version are being loaded from the `OvalRootType` but looks to me that it's never set.

parlt91 committed:
Fix code compilation

Signed-off-by: Pascal Arlt <parlt@suse.com>

parlt91 committed:
Add suggested changes

Signed-off-by: Pascal Arlt <parlt@suse.com>

parlt91 committed:
Move parse methods only used in testing to it's on test util class

Signed-off-by: Pascal Arlt <parlt@suse.com>

parlt91 committed:
Add oval definitions bulk size config value

Signed-off-by: Pascal Arlt <parlt@suse.com>

parlt91 reviewed (approved):
Tests are green and all remaining comments have been addressed. Approving

parlt91 commented:
@rjmateus I know it's been a while, but maybe you can do a quick scan of the PR and approve as well if you have the time?

rjmateus commented:
Sorry @parlt91 , the backlog pile is bug and ONLY reaxh this one now. The pr looks good and I glad its finally merged :)
